flink-tests/src/test/java/org/apache/flink/test/checkpointing/ChangelogPeriodicMaterializationITCase.java

```diff
@@ -0,0 +1,206 @@
+                            @Override
+                            protected void beforeElement(SourceContext<Integer> ctx)
+                                    throws Exception {
+                                if (getRuntimeContext().getAttemptNumber() == 0
+                                        && currentIndex == TOTAL_ELEMENTS / 4) {
+                                    waitWhile(
+                                            () -> {
+                                                if (completedCheckpointNum.get() <= 0) {
+                                                    return true;
+                                                }
+                                                List<Long> allMaterializationId =
+                                                        getAllMaterializationId(
+                                                                checkpointFile, jobID);
+                                                if (!allMaterializationId.isEmpty()) {
+                                                    currentMaterializationId
+                                                            .get()
+                                                            .addAll(allMaterializationId);
+                                                    Collections.sort(
+                                                            currentMaterializationId.get());
+                                                    currentCheckpointNum
+                                                            .get()
+                                                            .compareAndSet(
+                                                                    0,
+                                                                    completedCheckpointNum.get());
+                                                    return false;
+                                                }
+                                                return true;
+                                            });
+
+                                    throwArtificialFailure();
+                                } else if (getRuntimeContext().getAttemptNumber() == 1
+                                        && currentIndex == TOTAL_ELEMENTS / 2) {
+                                    waitWhile(
+                                            () -> {
+                                                if (completedCheckpointNum.get()
+                                                        <= currentCheckpointNum.get().get()) {
+                                                    return true;
+                                                }
+                                                List<Long> allMaterializationId =
+                                                        getAllMaterializationId(
+                                                                checkpointFile, jobID);
+                                                Collections.sort(allMaterializationId);
```

Comment:
Can't we use Set instead of sorting to check for equality?

Reply:
And probably it's better to use `StateHandleID` collected from the materialized states to avoid collisions between different subtasks if DoP > 1.